### ISSUE TITLE
MarginLoss produces very large loss if there are no active pairs

### ISSUE THREAD

**Issue by jac99** [KevinMusgrave/pytorch-metric-learning#98]
It may happen, that late in the training process both positive and negative distances in all triplets in some batch fall below thresholds. In such case MarginLoss produces very large loss (like 997187911680.0).  This breaks computation of batch statistics, such as a mean loss per batch, which I use in my code.

I looked into MarginLoss code and this is because when **pair_count (number of active pairs) becomes zero**,  total loss is equal to **beta_reg_loss divided by 1e-16**. Which produces a very large number. This is done by below piece of code:

```
        pair_count = self.num_pos_pairs + self.num_neg_pairs 

        return (torch.sum(pos_loss + neg_loss) + beta_reg_loss) / (pair_count + 1e-16)
```

When pair_count is zero, I think it's more logical to return loss equal to zero. Not some very large number. Maybe MarginLoss code could be amended to something like:

```
        return (torch.sum(pos_loss + neg_loss) + beta_reg_loss) / (max(pair_count,1))
```
This would prevent returning very large loss, when pair_count is zero. Or maybe more appropriate would be:
```
        if pair_count >= 1:
              return (torch.sum(pos_loss + neg_loss) + beta_reg_loss) / pair_count
       else:
              return ...grad enabled tensor set to zero (or something that doesn't break loss.backward()...
```

**Comment by KevinMusgrave** [comment]
The if/else looks good. We can just return 0 in the else statement, because the base class takes care of the loss.backward problem you mentioned.

**Comment by KevinMusgrave** [comment]
I made the changes you suggested, plus a couple others. See [this commit](https://github.com/KevinMusgrave/pytorch-metric-learning/commit/2dd61bb34c3da8b13c6ce974a5a67cae0c788483) to view the changes.

The changes are available in v0.9.86.dev3:
```
pip install pytorch-metric-learning==0.9.86.dev3
```

Let me know if it works

**Comment by jac99** [comment]
Hi Kevin, thanks for quick answer. The changes look good and I've re-run the code. Good idea to add margin_loss and beta_reg_loss to recordable attributes. This allows observing individual loss components.

However when pair_count >= 1, `margin_loss` and `beta_reg_loss` recordable attributes are one-element tensors. But when pair_count=0, `margin_loss` and `beta_reg_loss` recordable attributes are numbers (not tensors). This break my code to calculate per-batch statistics.
I added a simple fix in my code to handle both cases correctly and re-run the training loop. I'll let you known if everything finishes successfully.

Maybe it's worth to consider data type od recordable attributes? My feeling is, that these attributes are used for monitoring of the training process or displaying statistics about dynamics of the training process. So they should be all numbers not tensors. Currently, some recordable attributes are numbers (like num_zero_triplets in TripletMarginLoss) and some are grad-enabled tensors. I'm not sure if this is not causing some extra-memory consumption or other performance issues to keep track of computation graphs for each of these attributes.

**Comment by KevinMusgrave** [comment]
In my logging code I [detach if necessary](https://github.com/KevinMusgrave/record-keeper/blob/master/record_keeper/utils.py#L39). I suppose we could move some of that logic into the "recordable attributes" part of this library, though right now, "recordable attributes" is just a list of strings that correspond to object attributes.

**Comment by KevinMusgrave** [comment]
Closing this issue as its now in the latest version 0.9.86. I made a [separate issue](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/101) for the data types of recordable attributes, so we can continue the discussion there.
